### PR TITLE
doc: Add SDS (secret discovery service)

### PR DESCRIPTION
**Pull request by qiwzhang**
*Description*: Add SDS feature in the doc.
For https://github.com/envoyproxy/envoy/issues/4471

*Risk Level*: None

*Testing*: None

*Docs Changes*:
Yes, update the doc

*Release Notes*:
[Optional Fixes #Issue]
[Optional *Deprecated*:]

### CODE REVIEW

**Comment by qiwzhang** [comment]
@JimmyCYJ @lizan could you help to review it?

File: docs/root/intro/arch_overview/ssl.rst [hunk 1]
```diff
@@ -85,6 +85,126 @@ standard Debian installations. Common paths for system CA bundles on Linux and B
 See the reference for :ref:`UpstreamTlsContexts <envoy_api_msg_auth.UpstreamTlsContext>` and
 :ref:`DownstreamTlsContexts <envoy_api_msg_auth.DownstreamTlsContext>` for other TLS options.
 
+.. _secret_discovery_service:
```
Comment (lizan) [inline docs/root/intro/arch_overview/ssl.rst]: shall we put this in a new file like docs/root/configuration/secret.rst and put a link from here?

File: docs/root/intro/version_history.rst [hunk 2]
```diff
@@ -3,6 +3,8 @@ Version history
 
 1.8.0 (Pending)
 ===============
+* tls: implemented `SecretDiscoveryService <https://github.com/envoyproxy/envoy/blob/master/api/envoy/service/\
```
Comment (lizan) [inline docs/root/intro/version_history.rst]: alphabetical order.

File: docs/root/intro/arch_overview/ssl.rst [hunk 3]
```diff
@@ -85,6 +85,126 @@ standard Debian installations. Common paths for system CA bundles on Linux and B
...
+            tls_certificate_sds_secret_configs:
+            - name: "server_cert"
+              sds_config:
+                api_config_source:
+                  api_type: GRPC
+                  grpc_services:
+                  - google_grpc:
+                    target_uri: unix:/tmp/uds_path
+            validation_context_sds_secret_config:
+            - name: "validation_context"
```
Comment (wozz) [inline docs/root/intro/arch_overview/ssl.rst]: validation context isn't a list

File: docs/root/intro/arch_overview/ssl.rst (hunk 1 above)
Comment (qiwzhang) [reply docs/root/intro/arch_overview/ssl.rst]: Done.  move it into a separate file secret.rst under configuration.

File: docs/root/intro/version_history.rst (hunk 2 above)
Comment (qiwzhang) [reply docs/root/intro/version_history.rst]: Done

File: docs/root/intro/arch_overview/ssl.rst (hunk 3 above)
Comment (qiwzhang) [reply docs/root/intro/arch_overview/ssl.rst]: Done fixed it.  Thanks.

File: docs/root/configuration/secret.rst [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+==============================
+
+TLS certificates, the secrets, can be specified in the bootstrap.static_resource
+:ref:`secrets <envoy_api_field_config.bootstrap.v2.Bootstrap.StaticResources.secrets>`.
+But they can also be fetched remotely by secret discovery service (SDS).
+
+The benefits of remote fetch are
+
+* Easy to rotate expired certificates. SDS servers just push new certificates to Envoy.
+* Certificates are safe; they are not build into images, they can be fetched by secure channels such as Unix Domain Socket.
```
[the remark this answers is not in the document]
Comment (qiwzhang) [reply docs/root/configuration/secret.rst]: Are you referring to the certs that talks to SDS server?   In our cases, we are using Unix Domain Socket with the SDS server running in the same Node. It is safe, so not need certs for this call.

[the remark this answers is not in the document]
Comment (htuch) [reply docs/root/configuration/secret.rst]: I think we need to describe this in a more general way than assuming this feature only exists for Istio. Can you introduce the problem a bit more verbosely, describe UDS as a solution and maybe others? This is going to be a key operational issue. I realize we can't cover all possible deployments and environments, but more than one sentence would make sense.

[the remark this answers is not in the document]
Comment (qiwzhang) [reply docs/root/configuration/secret.rst]: Done.  Updated

[the remark this answers is not in the document]
Comment (ramaraochavali) [reply docs/root/configuration/secret.rst]: Yes. I was referring to securing SDS calls .  Thanks @htuch for explaining it more clearly.

File: docs/root/configuration/secret.rst [hunk 5]
```diff
@@ -7,10 +7,9 @@ TLS certificates, the secrets, can be specified in the bootstrap.static_resource
...
 But they can also be fetched remotely by secret discovery service (SDS).
 
-The benefits of remote fetch are
+The most important benefit of SDS is to simplify the certificate management. Without this feature, in k8s deployment, certificates must be created as secrets and mounted into the proxy containers. If certificates are expired, the secrets need to be updated and the proxy containers need to be re-deployed. With SDS, a certral SDS server will push certificates to all proxy instances. If certificates are expired, the server just pushes new certificates to Envoy instances, Envoy will use the new ones right away without re-deployment.
 
-* Easy to rotate expired certificates. SDS servers just push new certificates to Envoy.
-* Certificates are safe; they are not build into images, they can be fetched by secure channels such as Unix Domain Socket.
+The connection bewteeen Envoy proxy and SDS server has to be secure. One option is to run the SDS server in the same VM (node) and use Unix Domain Socket for the connection. Otherwise it requires mTLS between the proxy and SDS server. In this case, the client certificates to the SDS server has to be mounted into Envoy container.
```
Comment (htuch) [inline docs/root/configuration/secret.rst]: I would rephrase the bit about "mounted into Envoy container" as "In this case, the client certificates must be statically  configured.". This is a bit more general than referring to the specifics of Docker containers and file mounting.

File: docs/root/configuration/secret.rst [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+Secret discovery service (SDS)
+==============================
+
+TLS certificates, the secrets, can be specified in the bootstrap.static_resource
+:ref:`secrets <envoy_api_field_config.bootstrap.v2.Bootstrap.StaticResources.secrets>`.
+But they can also be fetched remotely by secret discovery service (SDS).
+
+The most important benefit of SDS is to simplify the certificate management. Without this feature, in k8s deployment, certificates must be created as secrets and mounted into the proxy containers. If certificates are expired, the secrets need to be updated and the proxy containers need to be re-deployed. With SDS, a certral SDS server will push certificates to all proxy instances. If certificates are expired, the server just pushes new certificates to Envoy instances, Envoy will use the new ones right away without re-deployment.
+
+The connection bewteeen Envoy proxy and SDS server has to be secure. One option is to run the SDS server in the same VM (node) and use Unix Domain Socket for the connection. Otherwise it requires mTLS between the proxy and SDS server. In this case, the client certificates to the SDS server has to be mounted into Envoy container.
```
[the remark this answers is not in the document]
Comment (qiwzhang) [reply docs/root/configuration/secret.rst]: Update the doc to describe how certificates are fetched during initialization.

File: docs/root/configuration/secret.rst (hunk 5 above)
Comment (qiwzhang) [reply docs/root/configuration/secret.rst]: Done

File: docs/root/configuration/secret.rst [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+If a listener server certificates needed to be fetched by SDS remotely, it will NOT be marked as active, its port will not be opened before the certificates are fetched. If Envoy fails to fetch the certificates due to connection failures, or bad response data, the listener will be marked as active, and the port will be open, but the connection to the port will be reset.
+
+Upstream cluster is handled in the similar way, if a cluster client certificates needed to be fetched by SDS remotely, it will NOT be marked as active and it will not be used before the certificates are fetched. If Envoy fails to fetch the certificates due to connection failures, or bad response data, the cluster will be marked as active, it can be used to handle the requests, but the requests routed to that cluster will be rejected.
+
+If a static cluster is using SDS, and it needs to define a SDS cluster (unless google grpc is used which doens't need a cluster), the SDS cluster has to be defined before the static cluster using it.
```
Comment (htuch) [inline docs/root/configuration/secret.rst]: Hmm, I'm a bit confused here. Are you talking about literal bootstrap cluster order? This seems a bit surprising to me that we have a definition order dependency here. In ClusterManager, we already have an notion of primary/secondary initialization order to deal with EDS cluster that depend on the EDS delivery cluster, how come we didn't extend this notion to SDS? CC @mattklein123

File: docs/root/configuration/secret.rst [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+The connection bewteeen Envoy proxy and SDS server has to be secure. One option is to run the SDS server in the same host and use Unix Domain Socket for the connection. Otherwise it requires mTLS between the proxy and SDS server. In this case, the client certificates for the SDS connection must be statically configured.
```
Comment (htuch) [inline docs/root/configuration/secret.rst]: "on the same host"

File: docs/root/configuration/secret.rst [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+SDS server
+----------
+
+A SDS server needs to implement the gRPC service `SecretDiscoveryService <https://github.com/envoyproxy/envoy/blob/master/api/envoy/service/discovery/v2/sds.proto>`_.
+It follows the same protocol as other `xDS <https://github.com/envoyproxy/data-plane-api/blob/master/XDS_PROTOCOL.md>`_
+
+SDS Configuration
+-----------------
+
+:ref:`SdsSecretConfig <envoy_api_msg_auth.SdsSecretConfig>` is used to specify the secret. Its field "name" is a required field. If its "sds_config" field is empty, the "name" field specifies the secret in the bootstrap static_resource :ref:`secrets <envoy_api_field_config.bootstrap.v2.Bootstrap.StaticResources.secrets>`. Otherwise, it specifies the SDS server as :ref:`ConfigSource <envoy_api_msg_core.ConfigSource>`. Only gRPC is supported for the SDS service so its "api_config_source" must specify a "grpc_service".
```
Comment (htuch) [inline docs/root/configuration/secret.rst]: Nit: when referring to fields, use `*name*` syntax for RST, etc. This is more consistent with the rest of the docs.

File: docs/root/configuration/secret.rst [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+SdsSecretConfig is used in two fields in :ref:`CommonTlsContext <envoy_api_msg_auth.CommonTlsContext>`. The first field is "tls_certificate_sds_secret_configs" to use SDS to get :ref:`TlsCertificate <envoy_api_msg_auth.TlsCertificate>`. The second field is "validation_context_sds_secret_config" to use SDS to get :ref:`CertificateValidationContext <envoy_api_msg_auth.CertificateValidationContext>`.
+
+Examples one: static_resource
+-----------------------------
+
+This example show how to configurate secrets in the static_resource:
```
Comment (htuch) [inline docs/root/configuration/secret.rst]: s/configurate/configure/

File: docs/root/configuration/secret.rst [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+
+.. code-block:: yaml
+
+  static_resources:
+    secrets:
+      - name: "server_cert"
```
Comment (htuch) [inline docs/root/configuration/secret.rst]: Nit: in YAML, you don't normally need `""` quotes unless for some disambiguation.

File: docs/root/configuration/secret.rst [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+              name: "validation_context"
+              sds_config:
+                api_config_source:
+                  api_type: GRPC
+                  grpc_services:
+                    envoy_grpc:
+                      cluster_name: sds_server_uds
+
+
+For illustration, above example uses 3 ways to access the SDS server. A gRPC SDS server can be reached by Unix Domain Socket path "/tmp/uds_path" and "127.0.0.1:8234" by mTLS. It can provide three secrets, "client_cert", "server_cert" and "validation_context". In the config, cluster "example_cluster" certificates "client_cert" is configured to use google_grpc by using UDS to talk to SDS server. The Listener needs to fetch "server_cert" and "validation_context" from the SDS server. The "server_cert" is using envoy_grpc with cluster "sds_server_mtls" configured with client certificates to use mTLS to talk to SDS server. The "validate_context" is using envoy_grpc with cluster "sds_server_uds" configured with UDS path to talk to the SDS server.
```
Comment (htuch) [inline docs/root/configuration/secret.rst]: Nit: can you cleanup the formatting of fields, values and gRPC types here to make it look pretty and consistent with the rest of Envoy docs? Thanks.

File: docs/root/configuration/secret.rst (hunk 7 above)
Comment (qiwzhang) [reply docs/root/configuration/secret.rst]: The current implementation of primary & secondary cluster initialization is simple;  only EDS is secondary, all other cluster types are primary.   In order to use this mechanism for SDS static clusters initialization, we have to mark any clusters using remote SDS as secondary.  

It is a TODO: created https://github.com/envoyproxy/envoy/issues/4522 to track this issue

File: docs/root/configuration/secret.rst (hunk 12 above)
Comment (qiwzhang) [reply docs/root/configuration/secret.rst]: Done

File: docs/root/configuration/secret.rst (hunk 11 above)
Comment (qiwzhang) [reply docs/root/configuration/secret.rst]: Done

File: docs/root/configuration/secret.rst (hunk 10 above)
Comment (qiwzhang) [reply docs/root/configuration/secret.rst]: Done

File: docs/root/configuration/secret.rst (hunk 9 above)
Comment (qiwzhang) [reply docs/root/configuration/secret.rst]: Done

File: docs/root/configuration/secret.rst (hunk 8 above)
Comment (qiwzhang) [reply docs/root/configuration/secret.rst]: Done

File: docs/root/configuration/secret.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (qiwzhang) [reply docs/root/configuration/secret.rst]: Done

File: docs/root/configuration/secret.rst [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+
+If a listener server certificates needed to be fetched by SDS remotely, it will NOT be marked as active, its port will not be opened before the certificates are fetched. If Envoy fails to fetch the certificates due to connection failures, or bad response data, the listener will be marked as active, and the port will be open, but the connection to the port will be reset.
+
+Upstream cluster is handled in the similar way, if a cluster client certificates needed to be fetched by SDS remotely, it will NOT be marked as active and it will not be used before the certificates are fetched. If Envoy fails to fetch the certificates due to connection failures, or bad response data, the cluster will be marked as active, it can be used to handle the requests, but the requests routed to that cluster will be rejected.
```
[the remark this answers is not in the document]
Comment (qiwzhang) [reply docs/root/configuration/secret.rst]: Done

File: docs/root/configuration/secret.rst [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+==============================
+
+TLS certificates, the secrets, can be specified in the bootstrap.static_resource
+:ref:`secrets <envoy_api_field_config.bootstrap.v2.Bootstrap.StaticResources.secrets>`.
+But they can also be fetched remotely by secret discovery service (SDS).
+
+The most important benefit of SDS is to simplify the certificate management. Without this feature, in k8s deployment, certificates must be created as secrets and mounted into the proxy containers. If certificates are expired, the secrets need to be updated and the proxy containers need to be re-deployed. With SDS, a certral SDS server will push certificates to all proxy instances. If certificates are expired, the server just pushes new certificates to Envoy instances, Envoy will use the new ones right away without re-deployment.
+
+If a listener server certificates needed to be fetched by SDS remotely, it will NOT be marked as active, its port will not be opened before the certificates are fetched. If Envoy fails to fetch the certificates due to connection failures, or bad response data, the listener will be marked as active, and the port will be open, but the connection to the port will be reset.
```
[the remark this answers is not in the document]
Comment (qiwzhang) [reply docs/root/configuration/secret.rst]: Done

File: docs/root/configuration/secret.rst [hunk 15]
```diff
@@ -0,0 +1,146 @@
+.. _config_secret_discovery_service:
+
+Secret discovery service (SDS)
+==============================
+
+TLS certificates, the secrets, can be specified in the bootstrap.static_resource
+:ref:`secrets <envoy_api_field_config.bootstrap.v2.Bootstrap.StaticResources.secrets>`.
+But they can also be fetched remotely by secret discovery service (SDS).
+
+The most important benefit of SDS is to simplify the certificate management. Without this feature, in k8s deployment, certificates must be created as secrets and mounted into the proxy containers. If certificates are expired, the secrets need to be updated and the proxy containers need to be re-deployed. With SDS, a certral SDS server will push certificates to all proxy instances. If certificates are expired, the server just pushes new certificates to Envoy instances, Envoy will use the new ones right away without re-deployment.
```
Comment (lizan) [inline docs/root/configuration/secret.rst]: s/certral/central/
s/proxy instances/Envoy instances/

Comment (qiwzhang) [reply docs/root/configuration/secret.rst]: Done

File: docs/root/configuration/secret.rst (hunk 7 above)
Comment (htuch) [reply docs/root/configuration/secret.rst]: @qiwzhang thanks for the explanation.

**Comment by mattklein123** [comment]
Just tagging that I want to read this over. Will do today. Please don't merge yet.
